File: app/src/main/java/net/veldor/flibustaloader/adapters/SubscribesAdapter.java
package net.veldor.flibustaloader.adapters;

import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import net.veldor.flibustaloader.App;
import net.veldor.flibustaloader.BR;
import net.veldor.flibustaloader.R;
import net.veldor.flibustaloader.databinding.SubscriptionItemBinding;
import net.veldor.flibustaloader.selections.SubscriptionItem;

import java.util.ArrayList;

public class SubscribesAdapter extends RecyclerView.Adapter<SubscribesAdapter.ViewHolder> {
    private ArrayList<SubscriptionItem> mItems;
    private LayoutInflater mLayoutInflater;

    public SubscribesAdapter(ArrayList<SubscriptionItem> itemsList) {
        mItems = itemsList;
    }

    @NonNull
    @Override
    public SubscribesAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        if (mLayoutInflater == null) {
            mLayoutInflater = LayoutInflater.from(viewGroup.getContext());
        }
        SubscriptionItemBinding binding = DataBindingUtil.inflate(mLayoutInflater, R.layout.subscription_item, viewGroup, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull SubscribesAdapter.ViewHolder viewHolder, int i) {
        viewHolder.bind(mItems.get(i));
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    public void changeList(ArrayList<SubscriptionItem> autocompleteValues) {
        mItems = autocompleteValues;
    }


    class ViewHolder extends RecyclerView.ViewHolder {

        private final ViewDataBinding mBinding;

        ViewHolder(ViewDataBinding binding) {
            super(binding.getRoot());
            mBinding = binding;
        }


        void bind(final SubscriptionItem item) {
            mBinding.setVariable(BR.subscriptions, item);
            mBinding.executePendingBindings();
            View container = mBinding.getRoot();
            View deleteBtn = container.findViewById(R.id.deleteItemBtn);
            if(deleteBtn != null){
                deleteBtn.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        switch (item.type) {
                            case "book":
                                App.getInstance().getBooksSubscribe().deleteValue(item.name);
                                break;
                            case "author":
                                App.getInstance().getAuthorsSubscribe().deleteValue(item.name);
                                break;
                            case "sequence":
                                App.getInstance().getSequencesSubscribe().deleteValue(item.name);
                                break;
                        }

                    }
                });
            }
        }
    }
}

File: app/src/main/java/net/veldor/flibustaloader/adapters/SubscribesAdapter.kt
package net.veldor.flibustaloader.adapters

import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import net.veldor.flibustaloader.App
import net.veldor.flibustaloader.BR
import net.veldor.flibustaloader.databinding.SubscriptionItemBinding
import net.veldor.flibustaloader.selections.SubscriptionItem
import net.veldor.flibustaloader.utils.SubscribeAuthors
import net.veldor.flibustaloader.utils.SubscribeBooks
import net.veldor.flibustaloader.utils.SubscribeSequences
import java.util.*

class SubscribesAdapter(private var mItems: ArrayList<SubscriptionItem>) :
    RecyclerView.Adapter<SubscribesAdapter.ViewHolder>() {
    private var mLayoutInflater: LayoutInflater =
        LayoutInflater.from(App.instance.applicationContext)

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ViewHolder {
        val binding = SubscriptionItemBinding.inflate(mLayoutInflater, viewGroup, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, i: Int) {
        viewHolder.bind(mItems[i])
    }

    override fun getItemCount(): Int {
        return mItems.size
    }

    fun itemAdded(item: SubscriptionItem?) {
        if (item != null) {
            // add item to top of list
            mItems.add(0, item)
            notifyItemInserted(0)
        }
    }

    fun itemRemoved(item: SubscriptionItem?) {
        if (item != null) {
            Log.d("surprise", "itemRemoved: deleting item")
            var foundedItem: SubscriptionItem? = null
            mItems.forEach {
                Log.d("surprise", "itemRemoved: ${it.name} ${it.type} ${item.name} ${item.type}")
                if (it.name == item.name && it.type == item.type) {
                    foundedItem = it
                }
            }
            if (foundedItem != null) {
                notifyItemRemoved(mItems.indexOf(foundedItem))
                mItems.remove(foundedItem)
            }
        }
    }

    fun changeList(autocompleteValues: ArrayList<SubscriptionItem>) {
        val prevSize = mItems.size
        notifyItemRangeRemoved(0, prevSize)
        mItems = autocompleteValues
        notifyItemRangeInserted(0, mItems.size)
    }

    class ViewHolder(private val mBinding: SubscriptionItemBinding) : RecyclerView.ViewHolder(
        mBinding.root
    ) {
        fun bind(item: SubscriptionItem) {
            mBinding.setVariable(BR.subscriptions, item)
            mBinding.executePendingBindings()
            mBinding.deleteItemBtn.setOnClickListener {
                when (item.type) {
                    "book" -> SubscribeBooks.instance.deleteValue(item.name)
                    "author" -> SubscribeAuthors.instance.deleteValue(item.name)
                    "sequence" -> SubscribeSequences.instance.deleteValue(item.name)
                }
            }
        }
    }
}